src/backend/optimizer/util/clauses.c

```diff
@@ -3276,8 +3276,16 @@ eval_const_expressions_mutator(Node *node,
 				{
 					Node	   *e;
 
-					e = eval_const_expressions_mutator((Node *) lfirst(arg),
-													   context);
+					if (sql_dialect == SQL_DIALECT_TSQL && !coalesceexpr->tsql_is_null)
```

Comment:
Please check with @surendravishnoi if this logic should be encapsulated in a hook. It looks fine to me but I wonder if this might lead to merge conflicts at time of community merges

Reply:
we can simply the condition put all checks in one line
if (not TSQL || tsql is null || type is coerce)
   eval const expr mutator

also add a comment explaining why we skip this line for given conditions

Reply:
We can not put all the checks in a single line as variable `e` would be NULL while checking the condition.

Reply:
I mean check like this
if (sql_dialext != SQL_DIALECT_TSQL || coalesceexpr->tsql_is_null || (lfirst(arg) == NULL) || (((Node *) lfirst(arg))->type ==  T_CoerceViaIO)
     e = eval_const_expressions_mutator((Node *) lfirst(arg), ...